### PR TITLE
Add Installer documentation

### LINKED ISSUE #226: Update Installer instructions in local-dev guide
We need to have instructions for using / modifying / testing the [installer library](https://github.com/vmware-tanzu/cluster-api-provider-bringyourownhost/tree/main/agent/installer). Add a new `Installer` section in [local-dev.md](https://github.com/vmware-tanzu/cluster-api-provider-bringyourownhost/blob/main/docs/local_dev.md)
This will help contributors to make bug fixes / enhancements to the installer library.

### PR DESCRIPTION
**Pull request by ratanasovvmw**
This PR fixes issue #226

Fixes #226

### CODE REVIEW

File: docs/local_dev.md [hunk 1]
```diff
@@ -179,3 +179,83 @@ kubectl delete cluster $CLUSTER_NAME
...
+Optional. This step describes downloading upstream k8s host components for Debian.
+```shell
+#Build docker image
+(cd agent/installer/bundle_builder/ingredients/deb/ && docker build -t byoh-ingredients-deb .)
+
+#Create a directory for the ingredients and download to it
+(mkdir -p byoh-ingredients-download && docker run --rm -v `pwd`/byoh-ingredients-download:/ingredients byoh-ingredients-deb)
+```
+### Downstream Ingredients
+This step describes providing custom k8s host components. They can be copied to to byoh-ingredients-download. Files must match the following globs:
```
Comment (anusha94) [inline docs/local_dev.md:197]:
```suggestion
This step describes providing custom kubernetes host components. They can be copied to `byoh-ingredients-download`. Files must match the following globs:
```

File: docs/local_dev.md [hunk 2]
```diff
@@ -179,3 +179,83 @@ kubectl delete cluster $CLUSTER_NAME
...
+### Upstream Ingredients
+Optional. This step describes downloading upstream k8s host components for Debian.
+```shell
+#Build docker image
+(cd agent/installer/bundle_builder/ingredients/deb/ && docker build -t byoh-ingredients-deb .)
+
+#Create a directory for the ingredients and download to it
+(mkdir -p byoh-ingredients-download && docker run --rm -v `pwd`/byoh-ingredients-download:/ingredients byoh-ingredients-deb)
+```
+### Downstream Ingredients
```
Comment (anusha94) [inline docs/local_dev.md:196]:
```suggestion
### Custom Ingredients
```

File: docs/local_dev.md [hunk 3]
```diff
@@ -179,3 +179,83 @@ kubectl delete cluster $CLUSTER_NAME
 docker rm -f $HOST_NAME
 kind delete cluster
 ```
+
+# Host Agent Installer
+The installer is responsible for detecting the BYOH OS, downloading a BYOH bundle and installing/uninstalling it.
+
+## Creating a BYOH Bundle
+### Upstream Ingredients
```
Comment (anusha94) [inline docs/local_dev.md:187]:
```suggestion
### Kubernetes Ingredients
```

Or some other better terminology than upstream

File: docs/local_dev.md [hunk 4]
```diff
@@ -179,3 +179,83 @@ kubectl delete cluster $CLUSTER_NAME
...
+Optional. This step describes downloading upstream k8s host components for Debian.
```
Comment (anusha94) [inline docs/local_dev.md:188]:
```suggestion
Optional. This step describes downloading kubernetes host components for Debian.
```

File: docs/local_dev.md [hunk 5]
```diff
@@ -179,3 +179,83 @@ kubectl delete cluster $CLUSTER_NAME
...
+*kubelet*.deb
+*kubectl*.deb
+*cri-tools*.deb
+*kubernetes-cni*.deb
+```
+
+## Building a BYOH Bundle
+```shell
+#Build docker image
+(cd agent/installer/bundle_builder/ && docker build -t byoh-build-push-bundle .)
```
Comment (anusha94) [inline docs/local_dev.md:210]:
```suggestion
cd agent/installer/bundle_builder/ && docker build -t byoh-build-push-bundle .
```

File: docs/local_dev.md [hunk 6]
```diff
@@ -179,3 +179,83 @@ kubectl delete cluster $CLUSTER_NAME
...
+```
+
+```shell
+#Build a BYOH bundle and publish it to an OCI-compliant repo
```
Comment (anusha94) [inline docs/local_dev.md:214]: Also worth adding how to push to a local registry? wdyt?

File: docs/local_dev.md [hunk 7]
```diff
@@ -179,3 +179,83 @@ kubectl delete cluster $CLUSTER_NAME
...
 ```
+
+# Host Agent Installer
+The installer is responsible for detecting the BYOH OS, downloading a BYOH bundle and installing/uninstalling it.
+
+## Creating a BYOH Bundle
+### Upstream Ingredients
+Optional. This step describes downloading upstream k8s host components for Debian.
+```shell
+#Build docker image
```
Comment (anusha94) [inline docs/local_dev.md:190]:
```suggestion
# Build docker image
```

File: docs/local_dev.md [hunk 8]
```diff
@@ -179,3 +179,83 @@ kubectl delete cluster $CLUSTER_NAME
...
+(cd agent/installer/bundle_builder/ingredients/deb/ && docker build -t byoh-ingredients-deb .)
+
+#Create a directory for the ingredients and download to it
```
Comment (anusha94) [inline docs/local_dev.md:193]:
```suggestion
# Create a directory for the ingredients and download to it
```
And all other occurrence of comments. Good to add a space for better readability

File: docs/local_dev.md [hunk 9]
```diff
@@ -179,3 +179,83 @@ kubectl delete cluster $CLUSTER_NAME
...
+
+```shell
+#Will detect the current OS and install BYOH bundle with k8s v1.22.3 from the default repo
+sudo ./cli --install --k8s v1.22.3
+```
+
+```shell
+#Will override the OS detection and will use custom repo
+sudo ./cli --install --os Ubuntu_20.04.1_x86-64 --bundle-repo 10.26.226.219:5000/repo --k8s v1.22.3
+```
```
Comment (anusha94) [inline docs/local_dev.md:261]: What about `uninstall`?

File: docs/local_dev.md [hunk 10]
```diff
@@ -179,3 +179,83 @@ kubectl delete cluster $CLUSTER_NAME
...
+#Will return the OS changes that installer will make during install and uninstall without actually doing them.
+./cli --preview-os-changes --os Ubuntu_20.04.*_x86-64 --k8s v1.22.3
+```
+
+```shell
+#Will detect the current OS and install BYOH bundle with k8s v1.22.3 from the default repo
+sudo ./cli --install --k8s v1.22.3
+```
+
+```shell
```
Comment (anusha94) [inline docs/local_dev.md:258]: Good to mention what the current default OS support is - either here or at the top in the Installer section.
P.S - yes, it needs to be updated everytime we add new support. I'm also okay having a support matrix in the top level `README`. Whatever you seem fit.

File: docs/local_dev.md (hunk 5 above)
Comment (ratanasovvmw) [reply docs/local_dev.md:210]: Braces are needed so the current dir of the current shell is not changed and you can execute all commands in sequence. Braces start a subshell.

File: docs/local_dev.md (hunk 6 above)
Comment (ratanasovvmw) [reply docs/local_dev.md:214]: Pushing to local registry should be like pushing to non-local.

If there is something very special in pushing, like we need, you can use the BUILD_ONLY=0 which is already explained.

**Review by anusha94 — APPROVED** [review]
LGTM

File: docs/local_dev.md [hunk 11]
```diff
@@ -179,3 +179,115 @@ kubectl delete cluster $CLUSTER_NAME
 docker rm -f $HOST_NAME
 kind delete cluster
 ```
+
+# Host Agent Installer
+The installer is responsible for detecting the BYOH OS, downloading a BYOH bundle and installing/uninstalling it.
```
Comment (dharmjit) [inline docs/local_dev.md:184]:
```suggestion
The installer is responsible for detecting the Host OS, downloading a BYOH bundle and installing/uninstalling it.
```

File: docs/local_dev.md [hunk 12]
```diff
@@ -179,3 +179,115 @@ kubectl delete cluster $CLUSTER_NAME
...
+    <tr>
+        <td>Ubuntu_20.04.*_x86-64</td>
+        <td>v1.22.3</td>
+        <td>byoh-bundle-ubuntu_20.04.1_x86-64_k8s_v1.22.3</td>
+    </tr>
+</table>
+The '*' in OS means that all Ubuntu 20.04 patches will be handled by this BYOH bundle.
+
+## Pre-requisites
+As of writing this, the following packages must be pre-installed on the BYOH host:
```
Comment (dharmjit) [inline docs/local_dev.md:207]: Question: These packages are only required for debian systems or other systems as well? Do we need to mention Ubuntu specifically?
```suggestion
As of writing this, the following packages must be pre-installed on the Host:
```

**Comment by lwmqwer** [comment]
lgtm too

**Review by lwmqwer — APPROVED** [review]

File: docs/local_dev.md [hunk 13]
```diff
@@ -179,3 +179,115 @@ kubectl delete cluster $CLUSTER_NAME
...
+
+## Creating a BYOH Bundle
+### Kubernetes Ingredients
+Optional. This step describes downloading kubernetes host components for Debian.
+```shell
+# Build docker image
+(cd agent/installer/bundle_builder/ingredients/deb/ && docker build -t byoh-ingredients-deb .)
+
+# Create a directory for the ingredients and download to it
+(mkdir -p byoh-ingredients-download && docker run --rm -v `pwd`/byoh-ingredients-download:/ingredients byoh-ingredients-deb)
```
Comment (dharmjit) [inline docs/local_dev.md:224]: Naive question: I was more aware of the usage `-v ${PWD}:/home/somepath` for volume mounts. But above is also okay if there are no downsides.

**Review by dharmjit — COMMENTED** [review]
Some Nits only

File: docs/local_dev.md (hunk 12 above)
Comment (ratanasovvmw) [reply docs/local_dev.md:207]: As of now we support only Ubuntu. For RHEL these will be rpms

File: docs/local_dev.md (hunk 13 above)
Comment (ratanasovvmw) [reply docs/local_dev.md:224]: I guess both are OK for an example.

**Pull request merged**